ui/panels-plugin/src/plugins/time-series-chart/utils/data-transform.ts

```diff
@@ -196,3 +195,27 @@ export function getThresholdSeries(
     },
   };
 }
+
+/**
+ * Converts percent threshold into absolute step value
+ * If max is undefined, use the max value from time series data as default
+ */
+export function convertPercentThreshold(percent: number, data: EChartsTimeSeries[], max?: number, min?: number) {
+  const percentDecimal = percent / 100;
+  const adjustedMax = max ?? findMax(data);
+  const adjustedMin = min ?? 0;
+  const total = adjustedMax - adjustedMin;
+  return percentDecimal * total + adjustedMin;
+}
+
+function findMax(timeSeries: EChartsTimeSeries[]) {
+  let max = 0;
+  timeSeries.forEach((series) => {
+    series.data.forEach((value: EChartsValues) => {
+      if (typeof value === 'number' && value > max) {
```

Comment:
We'll definitely want to test performance for some of the heavier queries internally since it's looping through all datapoints. Hopefully the useDeepMemo in TimeSeriesChartPanel helps enough too 🤞

Reply:
btw a query you can use occasionally for performance testing here that returns 1000+ series is:
```
rate(caddy_http_request_duration_seconds_bucket[$interval])
```
And in our product internally, queries from the cAdvisor dashboard return a ton of data